File: client_obsolete/RequestsAndResponses.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace VsSessionServer;

public enum NotificationType
{
    ProcessRestarted = 1,
    SessionTerminated = 2
}

public class VsSessionChangeNotification
{
    [Required]
    [JsonPropertyName("notification_type")]
    public NotificationType NotificationType { get; set; }

    [JsonPropertyName("pid")]
    public ushort PID { get; set; }

    [JsonPropertyName("session_id")]
    public string SessionId { get; set; } = string.Empty;

    public override string ToString()
    {
        var maybePID = this.PID != 0 ? $" (PID: {this.PID})" : string.Empty;
        return $"Session {this.SessionId}: {Enum.GetName(typeof(NotificationType), this.NotificationType)}{maybePID}";
    }
}

public class EnvVar
{
    [Required]
    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("value")]
    public string? Value { get; set; }
}

public class VsSessionRequest
{
    [Required]
    [JsonPropertyName("project_path")]
    public string ProjectPath { get; set; } = string.Empty;

    [Required]
    [JsonPropertyName("debug")]
    public bool Debug { get; set; }

    [JsonPropertyName("env")]
    public List<EnvVar> Environment { get; set; } = new List<EnvVar>();

    [JsonPropertyName("args")]
    public List<string> Arguments { get; set; } = new List<string>();
}

File: client-go/requests_and_responses.go
package main

import (
	"fmt"
)

type notificationType string

const (
	notificationTypeProcessRestarted  notificationType = "processRestarted"
	notificationTypeSessionTerminated notificationType = "sessionTerminated"
	notificationTypeServiceLogs       notificationType = "serviceLogs"
	notificationTypeProtected         notificationType = "protected"
)

type ideSessionNotificationBase struct {
	NotificationType notificationType `json:"notification_type"`
	SessionID        string           `json:"session_id,omitempty"`
}

func (snb *ideSessionNotificationBase) String() string {
	retval := fmt.Sprintf("Session %s: %s", snb.SessionID, snb.NotificationType)
	return retval
}

type ideRunSessionProcessChangedNotification struct {
	ideSessionNotificationBase
	PID int64 `json:"pid,omitempty"`
}

type ideRunSessionTerminatedNotification struct {
	ideRunSessionProcessChangedNotification
	ExitCode *int32 `json:"exit_code,omitempty"`
}

type ideSessionLogNotification struct {
	ideSessionNotificationBase
	IsStdErr   bool   `json:"is_std_err,omitempty"`
	LogMessage string `json:"log_message"`
}

type EnvVar struct {
	// Name of the environment variable
	Name string `json:"name"`

	// Value of the environment variable. Defaults to "" (empty string).
	// +optional
	Value string `json:"value,omitempty"`
	// CONSIDER allowing expansion of existing variable references e.g. using ${VAR_NAME} syntax and $$ to escape the $ sign
}

type VsSessionRequest struct {
	ProjectPath string   `json:"project_path"`
	Debug       bool     `json:"debug"`
	Env         []EnvVar `json:"env,omitempty"`
	Arguments   []string `json:"args,omitempty"`
}
